rootcacertpublisher: drop the namespace label from metrics to reduce its cardinality

stlaz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
The PR modifies the RootCA Publisher Controller metrics so that it does not include namespace as a label.

#### Which issue(s) this PR fixes:
The `root_ca_cert_publisher_sync_duration_seconds` metric tracks the sync duration in the root CA cert publisher per code and namespace. In clusters with a high namespace turnover (like CI clusters), this may cause the kube-controller-manager to expose over 100k series to Prometheus, which may cause degradation of that service.

Drop the `namespace` label to remove the metrics' cardinality, tracking this metric by namespace does not justify the impact of keeping it.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The `root_ca_cert_publisher_sync_duration_seconds` and `root_ca_cert_publisher_sync_total` metrics only track the response codes encountered but the namespace label was dropped.
 ```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

soltysh commented:
/ok-to-test

s-urbaniak commented:
/cc @ehashman @zshihang as well. context: https://github.com/kubernetes/kubernetes/pull/98731 caused quite some series churn in our clusters, so this is one suggestion to fix.

Alternative: The omission of tracking the `namespace` label is another alternative.

logicalhan commented:
/triage accepted 
/assign

logicalhan reviewed (commented):
We actually have a mechanism (depending on the version of kubernetes you are using) which can solve this problem without a code change. 

https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2305-metrics-cardinality-enforcement

You can pass in a flag to explicitly set cardinality boundaries for a metric and a given label.

logicalhan commented:
I would advocate for using the flag, since in normal cases, we actually do want to preserve the namespaces and their values. The CI case would be the exception to the rule.

s-urbaniak commented:
@logicalhan ah, thank you for the KEP reference, this is very helpful. The fact that this was exposed in a CI cluster is just an indicator that this will cause series churn in normal clusters as well, admittedly those, who leverage a lot of fluctuating namespaces.

The deletion of labels suggested in this PR does preserve previous data once scraped, it just doesn't expose them any more for future scrapes.

The way I read the KEP is that we would have to maintain an allowlist of namespaces we want to track with this metric, but then we'd lose information about other namespaces.

enj commented:
/remove-lifecycle stale

stlaz committed:
rootcacertpublisher: drop the namespace label from metrics to reduce its cardinality

The `root_ca_cert_publisher_sync_duration_seconds` metric tracks the sync
duration in the root CA cert publisher per code and namespace. In
clusters with a high namespace turnover (like CI clusters), this may
cause the kube-controller-manager to expose over 100k series to
Prometheus, which may cause degradation of that service.

Drop the `namespace` label to remove the metrics' cardinality, tracking
this metric by namespace does not justify the impact of keeping it.

stlaz commented:
Updated the metrics to no longer contain the "namespace" label as discussed on yesterday's sig/auth meeting.

liggitt commented:
/approve

would like lgtm from instrumentation reviewer as well

s-urbaniak commented:
/cc @logicalhan @zshihang @ehashman

mikedanese commented:
We talked about this in sig-auth and couldn't think of any value that the namespace label was adding to this metric.

logicalhan reviewed (commented):
/lgtm
/approve
